pkg/syncer/upsync/upsync_process.go

```diff
@@ -0,0 +1,113 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+func (c *Controller) process(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+
+	downstreamNamespace, downstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+
+	logger = logger.WithValues(logging.DownstreamNamespaceKey, downstreamNamespace, logging.DownstreamNameKey, downstreamName)
+
+	syncerInformer, ok := c.syncerInformers.InformerForResource(gvr)
+	if !ok {
+		return nil
+	}
+
+	// Upstream
+	var upstreamLocator *shared.NamespaceLocator
+	var locatorExists bool
+	// Namespace scoped resource
+	if downstreamNamespace != "" {
+		var nsObj runtime.Object
+		if nsObj, err = c.downstreamNamespaceLister.Get(downstreamNamespace); err != nil {
+			logger.Error(err, "Error getting downstream NS from lister")
+		}
+		nsMeta, ok := nsObj.(metav1.Object)
+		if !ok {
+			logger.Info("Error: downstream ns expected to be metav1.Object got", nsObj)
+		}
+
+		upstreamLocator, locatorExists, err = shared.LocatorFromAnnotations(nsMeta.GetAnnotations())
+		if err != nil {
+			logger.Error(err, "Error ")
+		}
+
+		if !locatorExists || upstreamLocator == nil {
+			return nil
+		}
+	}
+
+	var obj interface{}
+	var resourceExists bool
+	if obj, resourceExists, err = syncerInformer.DownstreamInformer.Informer().GetIndexer().GetByKey(key); err != nil {
+		return err
+	}
+	if downstreamNamespace == "" {
+		if !resourceExists {
+			return nil
+		}
+
+		objMeta, ok := obj.(metav1.Object)
+		if !ok {
+			logger.Info("Error: downstream cluster-wide res not of metav1 type")
+		}
+		if upstreamLocator, locatorExists, err = shared.LocatorFromAnnotations(objMeta.GetAnnotations()); err != nil {
+			logger.Error(err, "Error decoding annotations on DS cluster-wide res")
+		}
+
+		if !locatorExists || upstreamLocator == nil {
+			return nil
+		}
+	}
+
+	if upstreamLocator.SyncTarget.UID != c.syncTargetUID || upstreamLocator.SyncTarget.Workspace != c.syncTargetWorkspace.String() {
+		return nil
+	}
+
+	upstreamNamespace := upstreamLocator.Namespace
+	upstreamWorkspace := upstreamLocator.Workspace
+
+	if !resourceExists {
+		logger.Info("Downstream object doesn't exist")
+		return nil
+	}
+
+	u, ok := obj.(*unstructured.Unstructured)
+
+	if !ok {
+		// type mismatch
+		logger.Info("Type mismatch of resource object")
+		return nil
+	}
+	c.createResourceInUpstream(ctx, gvr, upstreamNamespace, upstreamWorkspace, u)
+
+	return nil
+}
+
+func (c *Controller) createResourceInUpstream(ctx context.Context, gvr schema.GroupVersionResource, upstreamNS string, upstreamLogicalCluster logicalcluster.Name, downstreamObj *unstructured.Unstructured) error {
+	// Make a deepcopy
+	downstreamObject := downstreamObj.DeepCopy()
+
+	// Create the resource
+	if _, err := c.upstreamClient.Cluster(upstreamLogicalCluster).Resource(gvr).Create(ctx, downstreamObject, metav1.CreateOptions{}); err != nil {
```

Comment:
Do you need to check to see if it's namespaced?

Reply:
Well, for now we'd upsync only cluster-wide resources (namely PVs), but in the future (as soon as we start upsyncing Pods related the synced deployments, we'll need to support namespaced resources.